### ISSUE TITLE
Remove ReferenceCounted.release(int) and retain(int)

### ISSUE THREAD

**Issue by normanmaurer** [netty/netty#8576]
Usually people just use `release()` or `retain()` but we also have overloads that take an `int` as argument. Having these overloads make the guards against overflow a lot harder and more complex. We should just remove these to simplify.

**Comment by normanmaurer** [comment]
FYI @ejona86 @carl-mastrangelo @njhill @trustin

**Comment by ejona86** [comment]
@normanmaurer, which guards are you talking about? I don't see much that would be simplified in `AbstractReferenceCounted`. Although maybe they don't exist because they would be too complicated.

**Comment by normanmaurer** [comment]
@ejona86 mostly related to overflow / underflow. Check 
https://github.com/netty/netty/issues/8563 for more details :)

**Comment by ejona86** [comment]
@normanmaurer, I didn't see anything special in #8563 that related to the `int` param. Everything there looked like it could happen even when using `1`.

**Comment by ejona86** [comment]
Limiting the `int` to `MAX_INT / 16` or even `100` doesn't seem like it would be a problem.

Note: I'm not really against this, but the rational didn't quite make sense to me. I'd agree the API is unlikely to be all that necessary.

**Comment by normanmaurer** [comment]
@ejona86 yeah... that said I think we should really just remove it if we do not need it.

**Comment by njhill** [comment]
My thinking exactly matches @ejona86's ... i.e. I'm not against this change but not sure that it will help "solve" the concurrency issues at hand apart from reducing the likelihood of race conditions occurring.

If retain uses an optimistic add, it's always possible for there to be N concurrent `retain(1)`s to cause the same problem as a `retain(N)` (though unlikely of course!)

**Comment by trustin** [comment]
A user might just want to release two or more times. Why should a user call `release()` n times when there is `release(n)`?

**Comment by trustin** [comment]
At least we should implement it as `for (int i = 0; i < n; i++) release();`

**Comment by njhill** [comment]
@trustin my example was just to illustrate how limiting the size can't in itself help to eliminate this kind of race condition, I wasn't suggesting it as an approach! If muli-release/retain is kept then I guess keeping it atomic would probably be better.
